Fix list absent

bstevenson opened the pull request:
### What does this PR do?

Adds a loop in the list_absent function to loop through the list of values as described in the documentation. 
### What issues does this PR fix or reference?

Bug #32052:  list_absent function doesn't loop through list of values
### Previous Behavior

List values never matched.

```
local:
----------
          ID: subrole
    Function: grains.list_absent
      Result: True
     Comment: Value ['item1', 'item2'] is absent from grain subrole
     Started: 16:58:25.333743
    Duration: 0.768 ms
     Changes:

Summary
------------
Succeeded: 1
Failed:    0
------------
```
### New Behavior

```
local:
----------
          ID: subrole
    Function: grains.list_absent
      Result: True
     Comment: Value item1 was deleted from grain subrole
              Value item2 was deleted from grain subrole
              Value itemx is absent from grain subrole
     Started: 01:03:58.924140
    Duration: 10048.179 ms
     Changes:
              ----------
              deleted:
                  - item1
                  - item2

Summary
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
```
### Tests written?

No

bstevenson committed:
Loop through list values in list_absent

Fixes #32052. List_absent now loops through items list.

bstevenson committed:
added whitespace

bstevenson committed:
cleaned up formating

bstevenson committed:
fix formating

bstevenson committed:
remove whitespace end of line 186:q

jfindlay commented on salt/states/grains.py:
```diff
@@ -169,26 +169,29 @@ def list_absent(name, value):
     '''
     ret = {'name': name,
-           'changes': {},
+           'changes': [],
```
@bstevenson, thanks for working on this.  `changes` needs to be a dictionary and `comment` needs to be a string as described in the [state docs](https://docs.saltstack.com/en/latest/ref/states/writing.html#return-data).

bstevenson committed:
Updated to conform to proper ret values

bstevenson committed:
Comments and Changes output fixes

Changes returns a dict with deleted key and list of values.
Improved newline handling for comment output

bstevenson committed:
Removed has_key in lieu of in

bstevenson committed:
Fixed negation logic

cachedout commented:
This has a failing test. @bstevenson could you please take a look? https://jenkins.saltstack.com/job/salt-pr-rs-ubuntu14.04-n/10898/

bstevenson commented:
@jfindlay I think the test needs to be modified to look for a list of grain values since list_absent now looks for a list of values.  The updates I made now always returns a list of values even if one grain value is passed.

```Traceback (most recent call last): File "/testing/tests/unit/states/grains_test.py", line 161, in test_list_absent ret1) AssertionError: {'comment': 'Value edam in grain cheese is set to be deleted', 'changes': {'dele [truncated]... != {'comment': 'Value edam in grain cheese is set to be deleted', 'changes': {'dele [truncated]... - {'changes': {'deleted': ['edam']}, ? - - + {'changes': {'deleted': 'edam'}, 'comment': 'Value edam in grain cheese is set to be deleted', 'name': 'cheese', 'result': None} ```

bstevenson commented:
@cachedout  Is it acceptable to modify existing tests?  I read https://docs.saltstack.com/en/latest/topics/development/tests/index.html but don't see a mention of modify tests, only writing new tests.

I believe lines 979 and 994 need to be changed from:
https://github.com/saltstack/salt/blob/develop/tests/unit/states/grains_test.py#L994
`self.assertEqual(ret['changes'], {'deleted': 'bar'})`
to
`self.assertEqual(ret['changes'], {'deleted': ['bar']})`

bstevenson committed:
unit test update

grain value looks for list in assertion

bstevenson committed:
set deleted value to list

bstevenson commented:
@cachedout Not sure the nature of the three remaining failures.  Let me know if you have any thoughts.

rallytime commented:
@bstevenson Those failures are not related. Thanks!
